plugins/inputs/kafka_consumer/kafka_consumer.go

```diff
@@ -241,11 +253,55 @@ func (k *KafkaConsumer) refreshTopics() error {
 	fingerprint := strings.Join(topicList, ";")
 	if fingerprint != k.fingerprint {
 		k.Log.Infof("updating topics: replacing %q with %q", k.allWantedTopics, topicList)
+		k.topicLock.Lock()
+		k.fingerprint = fingerprint
+		k.allWantedTopics = topicList
+		k.topicLock.Unlock()
+		// Having done that, we need to restart.
+		// This is tricky.  As soon as we call k.cancel() the old
+		// consumer group is no longer usable, so we need to get
+		// a new group and context ready and then pull a switcheroo
+		// quickly.
+		//
+		// I don't know if we lose messages between the cancel() and
+		// calling consumeTopics() on the new consumer group.
+		if k.consumer != nil {
+			k.Log.Info("restarting")
+			k.Log.Debug("creating new consumer group")
+			newConsumer, err := k.ConsumerCreator.Create(
+				k.Brokers,
+				k.ConsumerGroup,
+				k.config,
+			)
+			if err != nil {
+				acc.AddError(err)
+				return err
+			}
+			k.Log.Debug("getting new context before swapping consumers")
+			ctx, cancel := context.WithCancel(context.Background())
+			// Do we need to protect this with a lock?
+			// Do the switcheroo.
+			k.Log.Debug("replacing consumer group")
+			oldConsumer := k.consumer
+			k.consumer = newConsumer
+			k.cancel = cancel
+			// Do this in the background
+			go func() {
+				k.Log.Debug("closing old consumer group")
+				err = oldConsumer.Close()
+				if err != nil {
+					acc.AddError(err)
+					return
+				}
+			}()
+			k.Log.Info("starting new consumer group")
+			// Lock would end here.
+			k.consumeTopics(ctx, acc)
+			k.Log.Info("restarted with new topics")
+		}
```

Comment:
How about you do this in an own function and just return a `topicsChanged` boolean flag in`refreshTopics` ? This would avoid the intertwining between topic-refresh and consumer update...

Reply:
Refactored